### DISCUSSION TITLE
V1.01 Errata (Print version)

### DISCUSSION THREAD

**Discussion by wifinigel** [wifinigel/MikrotikScripting#5] [Errata]
The following issues have been found in version 1.01 of the print manuscript. These also apply to v1.0 (all updates should be applied automatically to the Kindle e-book version):

- Page 96: Section: Metacharacter: "?", second paragraph. `These strings will match this pattern:` should be `These strings will match the pattern "tre?"`
- Page 102: Section: code sample. `Perform logical AND` should be `Perform binary AND`
- Page 102: Section: Background, fourth. `In this example, we perform a logical [...]` should be `In this example, we perform a binary [...]`
- Page 103: e.g. { "apple", "orange", 5 }` should read `e.g. { "apple"; "orange"; 5 }`
- Page: 108, Section::find Command, first paragraph: `It reports the position of a sub-string within a string` should read: `It reports the position of the first occurrence of a sub-string within a string`
- Page: 118, Section: IP Addresses, second paragraph: `Any sequence of characters that contain a "." character (IPv4) or a ":" (IPv6) is considered to be an IP address. ` should read `Sequences of characters that contain "." (IPv4) or a ":" (IPv6) characters may be interpreted as  an IP address`. Also, add `Ensure type checking is used when using this data type before all operations.` as a final sentence to the same paragraph. 
- Page 144: Section: Global Commands, third paragraph: `used the ": put" command` should read `used the ":put" command` (space removed)
- Page: 149, Section: :execute, first paragraph: `execute { <commands> }` should read `execute { <commands> } [file=<file name>]`
- Page: 150, Section: rndnum (v7 only), last paragraph: `[admin@RouterOS] > :put [:rndnum from="100" to="200"];` should read `[admin@RouterOS] > :put [:rndnum from=100 to=200];` (remove quotes around numbers)
- Page: 152, Section: Joining Lines, first paragraph: wo instances of `running true` should read `running=yes` in code example. Fixed code example becomes:
```# filename: ch5-03-long-line.rsc

# Here is a long line
:put [/interface ethernet find where name~"ether" running=yes rx-fcs-error=0];

# Here is the same line across multiple lines
:put [/interface ethernet find where \
    name~"ether" \                                     
    running=yes \
    rx-fcs-error=0];
```
- Page 161: `:system` should read `/system` in code example
- Page 181: `Here's an analysis the script`. Needs fixing to read `Here's an analysis of the script`
- Page 191: In code example for ch6-12-global-vars3.rsc, command `:interface` should read `/interface`
- Page 192: In bullets 4 & 7, command `:interface` should read `/interface`
- Page 196: Section: What Are If Statements? Why Do We Use Them?, code example: `:system` should read `/system`
- Page 197: Section: What Are If Statements? Why Do We Use Them?, first paragraph: `:system` should read `/system`
- Page 201: code example: `:system` should read `/system`
- Page 202: code example: `:system` should read `/system`
- Page 203: code example: `:system` should read `/system`
- Page 204: code example: `:system` should read `/system`
- Page 206: code example: `:system` should read `/system`
- Page: 224: Section: While Loops, example code: `ch8-07-while-infinite.rsc` should read `ch8-06-while-infinite.rsc`
- Page: 224: Section: While Loops, example code: `delay 1` should read `:delay 1`
- Page 228: Two instances of `:fetch` should read: `/tool fetch`
- Page 238: First paragraph, `$i` should read $CharNumber`
- Page 246: 2nd from last paragraph, `/:environment print`  should read `:environment print`

**Comment by tabad** [comment]
Until now (I continue reading the book) I have found following issues in version 1.01 of the print edition. I can not confirm whether these erratas also apply to v1.0.

  * Page: 96, Section: *Metacharecter: "?"*, Paragraph: Second

    The pattern is missing. `These string will match this pattern:` should be  `These string will match the pattern "tre?":` or `These string will match this pattern: "tre?"`.

  * Page: 102, Section: *Background*, Paragraph: Forth

    The previous example does not show a *logical* AND operator but a *binary* AND, therefore `In this example, we perform a logical [...]` should be `In this example, we perform a binary [...]`. The first line of that example has the same mistake.

  * Page: 144, Section: *Global Commands*, Paragraph: Third

    Extra space. `[...] when we used the ": put" command [...]` should be `[...] when we used the ":put" command [...]`.

I keep reading and learning, and also *hunting* erratas. Thanks for your book Nigel.

**Reply by wifinigel** [reply]
All 3 items fixed and will appear in the v1.03 manuscript (all formats). Thanks @tabad !

**Comment by tabad** [comment]
I noticed that there is a script called `ch4-02-ip-prefix-slicer.rsc` (for RouterOS [v6](https://github.com/wifinigel/MikrotikScripting/blob/master/scripts/v6/ch4-02-ip-prefix-slicer.rsc) and [v7](https://github.com/wifinigel/MikrotikScripting/blob/master/scripts/v7/ch4-02-ip-prefix-slicer.rsc)) that is not mentioned in the printer version 1.01 of the book. In accordance with its name it should appears in chapter 4 but it uses a function, that is an advanced feature treated in chapter 9. Probably it should be mentioned inside section *IP Prefixes* of chapter 4.

**Reply by wifinigel** [reply]
Good spot! This file was included in error. I have removed it from the GitHub repo for now. I intended to use it in chapter 4, but decided it was too advanced for that point in the book, so did not include it in the manuscript. I forgot to remove the file :) I may include it in a future web-site article as it is a useful function. Thanks @tabad.

**Comment by tabad** [comment]
I have found the following new issues in version 1.01 of the print edition. As above, I can not confirm whether these erratas also apply to v1.0.

  * Page: 108, Section:*:find Command*, Paragraph: First

    Improve description/information. Since a sub-string can appears more than one time in the same string, the `:find` command reports the first occurrence of a sub-string within a string only, so the phrase `It reports the position of a sub-string within a string` could be more accurate like this: `It reports the position of the first occurrence of a sub-string within a string`.

  * Page: 149, Section: *:execute*, Paragraph: First

    Incomplete syntax. The complete syntax should be:

        execute script=<script name> [file=<file name>]
        execute { <commands> } [file=<file name>]

    The example of that page shows the use of the missing part of syntax.

  * Page: 149, Section: *:retry (v7 only)*, Paragraph: Last

    Sintax error. The syntax for `:retry` should be `:retry command={ <commands> } delay=<secs> [...]` instead of `:retry { <commands> } delay=<secs> [...]`

  * Page: 150, Section: *rndnum (v7 only)*, Paragraph: Last

    Syntax error. Values of parameters `from` and `to` of `:rndnum` command are numbers and should not be enclosed with double quotes. The example should be like this:

        [admin@RouterOS] > :put [:rndnum from=100 to=200];
        187
        [admin@RouterOS] >

  * Page: 152, Section: *Joining Lines*, Paragraph: First

    Error in conditional expression. I have tested the line `:put [/interface ethernet find where name~"ether" running true rx-fcs-error=0];` on a MikroTik CRS326-24G-2S+ with RouterOS v7.9 and it works.

        [admin@RouterOS] > :put [/interface ethernet find where name~"ether" running true rx-fcs-error=0];
        *18

    > **Note**: The conditional is equivalent to this other one: `[...] where name~"ether" and running=yes and true and rx-fcs-error=0`.

    But when using the same syntax I ask for the interfaces that are not running, the problem appears.

        [admin@RouterOS] > :put [/interface ethernet find where name~"ether" running false rx-fcs-error=0];
        
        [admin@RouterOS] > :put [/interface ethernet find where name~"ether" running=no rx-fcs-error=0];
        *1;*2;*3;*4;*5;*6;*7;*8;*9;*a;*b;*c;*d;*e;*f;*10;*11;*12;*13;*14;*15;*16;*17
        [admin@RouterOS] > :put [/interface ethernet find where name~"ether" !running rx-fcs-error=0];
        *1;*2;*3;*4;*5;*6;*7;*8;*9;*a;*b;*c;*d;*e;*f;*10;*11;*12;*13;*14;*15;*16;*17

    Any of the following forms would be a better approach to the problem.

        [admin@RouterOS] > :put [/interface ethernet find where name~"ether" and running=yes and rx-fcs-error=0];
        *18
        [admin@RouterOS] > :put [/interface ethernet find where name~"ether" and running and rx-fcs-error=0];
        *18
        [admin@RouterOS] > :put [/interface ethernet find where name~"ether" running rx-fcs-error=0];
        *18
        [admin@RouterOS] > :put [/interface ethernet find name~"ether" running rx-fcs-error=0];
        *18

    With multiple lines...

        [admin@RouterOS] > :put [/interface ethernet \
        [\... find where         \
        [\...     name~"ether"   \
        [\...     running        \
        [\...     rx-fcs-error=0 \
        [\... ];
        *18

  * Page: 192, Section: *Best Practice Tips When Using Global Variables*, Paragraphs: Seventh and Tenth

    Bad configuration command reference. In the seventh paragraph the reference to `":interface ethernet find"` should be changed for `"/:interface ethernet find"`. In the tenth paragraph the reference to `[:interface ethernet get $InterfaceId]->"running"` should be changed to `[/interface ethernet get $InterfaceId]->"running"`. The example `ch6-12-global-vars3.src` on the previous page also has this last error (issue #6).

**Reply by wifinigel** [reply]
All items fixed and will appear in the v1.03 manuscript (all formats). Thanks @tabad !

**Comment by tabad** [comment]
I have found the following new issues in version 1.01 of the print edition. As above, I can not confirm whether these erratas also apply to v1.0.

  * Page: 118, Section: *IP Adresses*, Paragraph: Second

    False statement. Any sequence of characters that contain a `.` character or a `:` character **is not always** considered an IP address. These are some examples:

        [admin@RouterOS] > :put [:typeof 00];
        num
        [admin@RouterOS] > :put [:typeof 00:];
        time
        [admin@RouterOS] > :put [:typeof 00:00];
        time
        [admin@RouterOS] > :put [:typeof 00:00:00];
        time
        [admin@RouterOS] > :put [:typeof 00:00:00:00];
        str
        [admin@RouterOS] > :put [:typeof 00:00:00:00:00];
        str
        [admin@RouterOS] > :put [:typeof 00:00:00:00:00:00];
        str
        [admin@RouterOS] > :put [:typeof 00:00:00:00:00:00:00];
        str

        [admin@RouterOS] > :put [:typeof 0];
        num
        [admin@RouterOS] > :put [:typeof 0.];
        str
        [admin@RouterOS] > :put [:typeof 0.0];
        ip
        [admin@RouterOS] > :put [:typeof 0.0.0];
        ip
        [admin@RouterOS] > :put [:typeof 0.0.0.0];
        ip
        [admin@RouterOS] > :put [:typeof 0.0.0.0.0];
        str

**Reply by wifinigel** [reply]
Yes, your observation is correct. 

I have generalized the statement to: "_Sequences of characters that contain "." (IPv4) or  ":" (IPv6) characters may be interpreted an IP address._". 

Having tried many different combinations, it is difficult to safely provided definitive formatting guidelines for all cases. I have added to a note to advise readers to use data type checking for this data type: "_Ensure type checking is used when using this data type before all operations_"

**Comment by tabad** [comment]
I have found the following new issues in version 1.01 of the print edition. As above, I can not confirm whether these erratas also apply to v1.0.

  * Page: 197, Section: *What Are If Statement? Why Do We Use Them?*, Paragraph: First

    Bad configuration command reference. In the first paragraph the reference to `[:system clock get time]` should be changed to `[/system clock get time]`. The example `ch7-01-basic-if.src` on the previous page also has this last error (issue #7).

**Reply by wifinigel** [reply]
Both items fixed and will appear in the v1.03 manuscript (all formats). Thanks @tabad !

**Comment by tabad** [comment]
I have found the following new issues in version 1.01 of the print edition. As above, I can not confirm whether these erratas also apply to v1.0.

  * Page: 224, Section: *While Loops*, Example Code

    Bad file name. The first line of that example code says it is the file `ch8-07-while-infinite.rsc` but it is actually the file `ch8-06-while-infinite.rsc`.

**Reply by wifinigel** [reply]
Item fixed and will appear in the v1.03 manuscript (all formats). Thanks @tabad !

**Comment by tabad** [comment]
I have found the following new issues in version 1.01 of the print edition. As above, I can not confirm whether these erratas also apply to v1.0.

  * Page: 245, Section: *Background*, First Line of Sample Code

    Extra `:`. The sample code

        :# clear global variable (function name)
        :set PrintInterfaceAttrsFunc;

should be

        # clear global variable (function name)
        :set PrintInterfaceAttrsFunc;

**Reply by wifinigel** [reply]
Item fixed and will appear in the v1.04 manuscript (all formats). Thanks @tabad !

**Comment by tabad** [comment]
I have found the following new issues in version 1.01 of the print edition. As above, I can not confirm whether these erratas also apply to v1.0.

  * Page: 271, Section: *Debugging*, First Sample Code

    Extra Spaces (Cosmetic). The sample code

        # debug function
        :global DebugFunc do={
              :global DEBUG;
              :if ($DEBUG) do={ :put "**Debug: $1";       }
        }

    should be

        # debug function
        :global DebugFunc do={
              :global DEBUG;
              :if ($DEBUG) do={ :put "**Debug: $1"; }
        }

    It's like it appears in `ch10-08-bad-script.rsc` file.

  * Page: 272, Section: *Profiling Code*, Last Sample Code

    Extra Spaces (Cosmetic). The sample code

        [admin@Router] > :put [:time {/import ch10-09-simple-pause.rsc }];
        Hello we're going to pause now...
        [...]

    should be

        [admin@Router] > :put [:time {/import ch10-09-simple-pause.rsc}];
        Hello we're going to pause now...
        [...]

  * Page: 273, Section: *Profiling Code*, Last Sample Code

    Extra Spaces (Cosmetic). The sample code

        [admin@Router] > :put [:time { /import ch10-10-add-remove-vlans.rsc }];
        
        Script file loaded and executed successfully
        [...]

    should be

        [admin@Router] > :put [:time {/import ch10-10-add-remove-vlans.rsc}];
        
        Script file loaded and executed successfully
        [...]

  * Page: 274, Section: *Profiling Code*, Last Sample Code

    Extra Spaces (Cosmetic) & File Name Error. The sample code

        [admin@Router] > :put [:time { /import ch10-011-add-remove-vlans.rsc }];
        
        Script file loaded and executed successfully
        [...]

    should be

        [admin@Router] > :put [:time {/import ch10-11-add-remove-vlans.rsc}];
        
        Script file loaded and executed successfully
        [...]

    Also note the change in the file name. The `ch10-011-add-remove-vlans.rsc` file does not exist.

**Reply by wifinigel** [reply]
Item fixed and will appear in the v1.04 manuscript (all formats). Thanks @tabad !
